File: ciel/evolution_hydra/LeaderNetwork.ts
/**
 * ═══════════════════════════════════════════════════════════════
 * LEADER-NETWORK — EventBus pour la communication inter-Olympiens
 * ═══════════════════════════════════════════════════════════════
 *
 * PHASE 4 : refactorisé pour exposer un canal de communication
 * sémiotique (tokens émergents) en plus du canal vectoriel brut.
 *
 * Canaux disponibles :
 *  - 'discovery'          : découverte partagée
 *  - 'dm_<toFactionId>'   : message direct
 *  - 'council_order'      : ordre du Conseil (Overseer)
 *  - 'vector_exchange'    : tenseur brut (legacy, ultra-efficace)
 *  - 'token'              : NOUVEAU — token émergent (EmergentLanguage)
 *
 * Théorie :
 *  - Le canal vectoriel transporte des nombres arbitraires
 *  - Le canal token transporte du sens auto-organisé
 *  - Les deux sont complémentaires : tokens pour le sens,
 *    vecteurs pour la haute fréquence
 */

import { EventEmitter } from 'events';
import logger from '../utils/logger.js';

export interface EmergentToken {
  /** Le token composé (concaténation de primitives). */
  symbol: string;
  /** Les intentions qui ont composé le token. */
  intents: Array<{
    domain: string;
    intensity: number;
    context: string;
    phi?: number;
  }>;
  /** Valence émotionnelle (-1 = menace, 0 = neutre, +1 = opportunité). */
  valence: number;
  /** Epoch ms. */
  timestamp: number;
  /** ID de l'agent émetteur. */
  originatorId: string;
  /** ID unique du token (pour déduplication). */
  id: string;
}

class NetworkManager extends EventEmitter {
  constructor() {
    super();
    // Allow up to 100 concurrent leaders/listeners
    this.setMaxListeners(100);
  }

  // A Leader broadcasts a discovery to all other leaders and the Overseer
  broadcastDiscovery(factionId: string, title: string, discovery: string): void {
    logger.debug(`[Leader Network] 📡 ${title} broadcasts: "${discovery.substring(0, 50)}..."`);
    this.emit('discovery', { factionId, title, discovery });
  }

  // A Leader wants to discuss something specifically with another
  sendDirectMessage(fromFactionId: string, toFactionId: string, message: string): void {
    this.emit(`dm_${toFactionId}`, { fromFactionId, message });
  }

  // The Overseer (Council) sends an inescapable order to all Leaders
  transmitCouncilOrder(order: string): void {
    logger.debug(`[Council Command] ⚠️ OVERSEER TRANSMITS DIRECTIVE: ${order}`);
    this.emit('council_order', order);
  }

  // --- OMNISCIENCE: Emergent Communication ---
  // A Leader or Clone sends a RAW vector signal (tensor) to another.
  // This bypasses natural language for ultra-efficient coordination.
  broadcastVector(fromId: string, toId: string, vector: number[]): void {
    this.emit('vector_exchange', { fromId, toId, vector });
    logger.debug(`[Emergent Comm] Vector Exchange: ${fromId} ➔ ${toId} (len: ${vector.length})`);
  }

  /**
   * PHASE 4 : diffuse un token émergent (sémiotique) sur le réseau.
   * Contrairement à broadcastVector, le token est du SENS — pas juste
   * un tenseur. Les listeners peuvent l'interpréter via EmergentLanguage.
   */
  broadcastToken(token: EmergentToken): void {
    logger.debug(`[Leader Network] 🗣️ Token: ${token.symbol} from ${token.originatorId} (valence: ${token.valence.toFixed(2)})`);
    this.emit('token', token);
  }

  /**
   * Variante "broadcast" ciblé : envoie un token à un agent spécifique.
   * Utilise le canal direct.
   */
  sendTokenTo(toId: string, token: EmergentToken): void {
    logger.debug(`[Leader Network] 📨 Token to ${toId}: ${token.symbol}`);
    this.emit(`token_${toId}`, token);
  }
}

export const LeaderNetwork = new NetworkManager();

File: ciel/evolution/leader_network.py
"""
LeaderNetwork - Inter-Agent Communication Channel
EventEmitter for faction coordination and emergent tokens.
"""
from __future__ import annotations

import logging
from dataclasses import dataclass, field
from typing import Any, Callable

logger = logging.getLogger(__name__)


@dataclass(slots=True)
class EmergentToken:
    """Self-organized semantic token"""
    symbol: str
    intents: list[dict[str, Any]]
    valence: float  # -1 = threat, 0 = neutral, +1 = opportunity
    timestamp: float
    originator_id: str
    id: str


@dataclass(slots=True)
class LeaderNetwork:
    """
    Leader Network - EventBus for Inter-Agent Communication
    Supports both vector and semantic token channels.
    """
    listeners: dict[str, list[Callable]] = field(default_factory=dict)
    max_listeners: int = 100

    def __post_init__(self) -> None:
        """Initialize network"""
        logger.info("[Leader Network] 📡 Leader Network initialized.")

    def broadcast_discovery(self, faction_id: str, title: str, discovery: str) -> None:
        """Broadcast discovery to all leaders"""
        logger.debug(
            f"[Leader Network] 📡 {title} broadcasts: \"{discovery[:50]}...\""
        )
        self._emit("discovery", {
            "faction_id": faction_id,
            "title": title,
            "discovery": discovery
        })

    def send_direct_message(
        self, from_faction_id: str, to_faction_id: str, message: str
    ) -> None:
        """Send direct message between factions"""
        channel = f"dm_{to_faction_id}"
        self._emit(channel, {
            "from_faction_id": from_faction_id,
            "message": message
        })

    def transmit_council_order(self, order: str) -> None:
        """Transmit directive from Council/Overseer"""
        logger.debug(
            f"[Council Command] ⚠️ OVERSEER TRANSMITS DIRECTIVE: {order}"
        )
        self._emit("council_order", order)

    def broadcast_vector(
        self, from_id: str, to_id: str, vector: list[float]
    ) -> None:
        """Broadcast raw vector (ultra-efficient coordination)"""
        self._emit("vector_exchange", {
            "from_id": from_id,
            "to_id": to_id,
            "vector": vector
        })
        logger.debug(
            f"[Emergent Comm] Vector Exchange: {from_id} ➔ {to_id} (len: {len(vector)})"
        )

    def broadcast_token(self, token: EmergentToken) -> None:
        """Broadcast emergent semantic token"""
        logger.debug(
            f"[Leader Network] 🗣️ Token: {token.symbol} from {token.originator_id} "
            f"(valence: {token.valence:.2f})"
        )
        self._emit("token", token)

    def send_token_to(self, to_id: str, token: EmergentToken) -> None:
        """Send token to specific agent"""
        logger.debug(
            f"[Leader Network] 📨 Token to {to_id}: {token.symbol}"
        )
        self._emit(f"token_{to_id}", token)

    def on(self, event: str, callback: Callable) -> None:
        """Register listener for event"""
        if event not in self.listeners:
            self.listeners[event] = []
        if len(self.listeners[event]) < self.max_listeners:
            self.listeners[event].append(callback)

    def off(self, event: str, callback: Callable) -> None:
        """Unregister listener"""
        if event in self.listeners:
            self.listeners[event] = [
                c for c in self.listeners[event] if c != callback
            ]

    def _emit(self, event: str, data: Any) -> None:
        """Emit event to all listeners"""
        if event in self.listeners:
            for callback in self.listeners[event]:
                try:
                    callback(data)
                except Exception as e:
                    logger.error(f"[Leader Network] Error in callback: {e}")

    def process(self, input_data: Any) -> dict[str, Any]:
        """
        Process network request.
        CIEL compatibility method.
        """
        return {
            "channels": len(self.listeners),
            "status": "active"
        }


# Global singleton instance
_leader_network: LeaderNetwork | None = None


def get_leader_network() -> LeaderNetwork:
    """Get or create global leader network"""
    global _leader_network
    if _leader_network is None:
        _leader_network = LeaderNetwork()
    return _leader_network
